In [ ]:
import pandas as pd
import numpy as np
import datetime


In [ ]:
df = pd.read_csv('train.csv')

In [ ]:
df.shape

(5847, 14)

In [ ]:
df


Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
0                 41000    Diesel       Manual      First  19.67 kmpl   
1                 46000    Petrol       Manual      First    13 km/kg   
2                 87000    Diesel       Manual      First  20.77 kmpl   
3                 40670    Diesel    Automatic     Second   15.2 kmpl   
4                 86999    Diesel       Manual      First  23.08 kmpl   
...                 ...       ...          ...        ...         ...   
5842              27365    Diesel       Manual      First   28.4 kmpl   
5843             100000    Diesel       Manual      First   24.4 kmpl   
5844              55000    Diesel       Manual     Second   14.0 kmpl   
5845              46000    Petrol       Manual      First   18.9 kmpl   
5846              47000    Diesel       Manual      First  25.44 kmpl   

       Engine      Power  Seats  New_Price  Price  
0     1582 CC  126.2 bhp    5.0        NaN  12.50  
1     1199 CC   88.7 bhp    5.0  8.61 Lakh   4.50  
2     1248 CC  88.76 bhp    7.0        NaN   6.00  
3     1968 CC  140.8 bhp    5.0        NaN  17.74  
4     1461 CC   63.1 bhp    5.0        NaN   3.50  
...       ...        ...    ...        ...    ...  
5842  1248 CC     74 bhp    5.0  7.88 Lakh   4.75  
5843  1120 CC     71 bhp    5.0        NaN   4.00  
5844  2498 CC    112 bhp    8.0        NaN   2.90  
5845   998 CC   67.1 bhp    5.0        NaN   2.65  
5846   936 CC   57.6 bhp    5.0        NaN   2.50  

[5847 rows x 14 columns]

In [ ]:
df.count()

Unnamed: 0           5847
Name                 5847
Location             5847
Year                 5847
Kilometers_Driven    5847
Fuel_Type            5847
Transmission         5847
Owner_Type           5847
Mileage              5845
Engine               5811
Power                5811
Seats                5809
New_Price             815
Price                5847
dtype: int64

In [ ]:
# Initialize empty lists to store statistics
columns = []
unique_values = []
missing_percentage = []
data_types = []

# Loop through the columns
for col in df.columns:
    columns.append(col)
    unique_values.append(df[col].nunique())
    missing_percentage.append(df[col].isna().sum()/ len(df))
    data_types.append(df[col].dtype)

# Create a summary DataFrame
summary_df = pd.DataFrame({
    'Column': columns,
    'Unique Values': unique_values,
    'Missing Percentage': missing_percentage,
    'Data Type': data_types
})

print(summary_df)

               Column  Unique Values  Missing Percentage Data Type
0          Unnamed: 0           5847            0.000000     int64
1                Name           1804            0.000000    object
2            Location             11            0.000000    object
3                Year             22            0.000000     int64
4   Kilometers_Driven           3019            0.000000     int64
5           Fuel_Type              3            0.000000    object
6        Transmission              2            0.000000    object
7          Owner_Type              4            0.000000    object
8             Mileage            419            0.000342    object
9              Engine            140            0.006157    object
10              Power            361            0.006157    object
11              Seats              8            0.006499   float64
12          New_Price            534            0.860612    object
13              Price           1366            0.000000   flo

# Dropping the columns

Dropping the column "Unnamed" since it looks like a counter and not providing the exact count of rows. It is counting the rows as 6018 when the real count is 5847 and it is not useful for analysis.

Dropped the column "New_Price" since there is not enough data in it. It has about 86% null values though it won't be very helpful by imputing (replacing with mean, mode). So at this point dropping it is a good option.


In [ ]:
drop = ["Unnamed: 0" , "New_Price"]
df = df.drop(columns= drop)

In [ ]:
df.count()


Name                 5847
Location             5847
Year                 5847
Kilometers_Driven    5847
Fuel_Type            5847
Transmission         5847
Owner_Type           5847
Mileage              5845
Engine               5811
Power                5811
Seats                5809
Price                5847
dtype: int64

In [ ]:
df['Engine'] = pd.to_numeric(df['Engine'].str.replace(r'[^\d.]', '', regex=True), errors='coerce')
df['Power'] = pd.to_numeric(df['Power'].str.replace(r'[^\d.]', '', regex=True), errors='coerce')

In [ ]:
df.fillna(df.mean(), inplace=True)

<ipython-input-36-bf934754a5b1>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.fillna(df.mean(), inplace=True)


In [ ]:
df.count()

Name                 5847
Location             5847
Year                 5847
Kilometers_Driven    5847
Fuel_Type            5847
Transmission         5847
Owner_Type           5847
Mileage              5847
Engine               5847
Power                5847
Seats                5847
Price                5847
dtype: int64

In [ ]:
def convert_to_float(value):
    if isinstance(value, str):
        value = value.replace(r'[^\d.]', '', regex=True)
        return float(value)
    return float(value)

# Apply the custom function to the 'Mileage' column
df['Mileage'] = df['Mileage'].apply(convert_to_float)
df['Engine'] = df['Engine'].apply(convert_to_float)
df['Power'] = df['Power'].apply(convert_to_float)


In [ ]:
df

Name    Location  Year  Kilometers_Driven  \
0     Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                         Honda Jazz V     Chennai  2011              46000   
2                    Maruti Ertiga VDI     Chennai  2012              87000   
3      Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4               Nissan Micra Diesel XV      Jaipur  2013              86999   
...                                ...         ...   ...                ...   
5842                  Maruti Swift VDI       Delhi  2014              27365   
5843          Hyundai Xcent 1.1 CRDi S      Jaipur  2015             100000   
5844             Mahindra Xylo D4 BSIV      Jaipur  2012              55000   
5845                Maruti Wagon R VXI     Kolkata  2013              46000   
5846             Chevrolet Beat Diesel   Hyderabad  2011              47000   

     Fuel_Type Transmission Owner_Type  Mileage  Engine   Power  Seats  Price  
0       Diesel       Manual      First    19.67  1582.0  126.20    5.0  12.50  
1       Petrol       Manual      First    13.00  1199.0   88.70    5.0   4.50  
2       Diesel       Manual      First    20.77  1248.0   88.76    7.0   6.00  
3       Diesel    Automatic     Second    15.20  1968.0  140.80    5.0  17.74  
4       Diesel       Manual      First    23.08  1461.0   63.10    5.0   3.50  
...        ...          ...        ...      ...     ...     ...    ...    ...  
5842    Diesel       Manual      First    28.40  1248.0   74.00    5.0   4.75  
5843    Diesel       Manual      First    24.40  1120.0   71.00    5.0   4.00  
5844    Diesel       Manual     Second    14.00  2498.0  112.00    8.0   2.90  
5845    Petrol       Manual      First    18.90   998.0   67.10    5.0   2.65  
5846    Diesel       Manual      First    25.44   936.0   57.60    5.0   2.50  

[5847 rows x 12 columns]

For one hot encoding pandas dummie function will create indicators for all catagories

In [ ]:
df = pd.get_dummies(df, columns=['Fuel_Type', 'Transmission'])

In [ ]:
df

Name    Location  Year  Kilometers_Driven  \
0     Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                         Honda Jazz V     Chennai  2011              46000   
2                    Maruti Ertiga VDI     Chennai  2012              87000   
3      Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4               Nissan Micra Diesel XV      Jaipur  2013              86999   
...                                ...         ...   ...                ...   
5842                  Maruti Swift VDI       Delhi  2014              27365   
5843          Hyundai Xcent 1.1 CRDi S      Jaipur  2015             100000   
5844             Mahindra Xylo D4 BSIV      Jaipur  2012              55000   
5845                Maruti Wagon R VXI     Kolkata  2013              46000   
5846             Chevrolet Beat Diesel   Hyderabad  2011              47000   

     Owner_Type  Mileage  Engine   Power  Seats  Price  Fuel_Type_Diesel  \
0         First    19.67  1582.0  126.20    5.0  12.50                 1   
1         First    13.00  1199.0   88.70    5.0   4.50                 0   
2         First    20.77  1248.0   88.76    7.0   6.00                 1   
3        Second    15.20  1968.0  140.80    5.0  17.74                 1   
4         First    23.08  1461.0   63.10    5.0   3.50                 1   
...         ...      ...     ...     ...    ...    ...               ...   
5842      First    28.40  1248.0   74.00    5.0   4.75                 1   
5843      First    24.40  1120.0   71.00    5.0   4.00                 1   
5844     Second    14.00  2498.0  112.00    8.0   2.90                 1   
5845      First    18.90   998.0   67.10    5.0   2.65                 0   
5846      First    25.44   936.0   57.60    5.0   2.50                 1   

      Fuel_Type_Electric  Fuel_Type_Petrol  Transmission_Automatic  \
0                      0                 0                       0   
1                      0                 1                       0   
2                      0                 0                       0   
3                      0                 0                       1   
4                      0                 0                       0   
...                  ...               ...                     ...   
5842                   0                 0                       0   
5843                   0                 0                       0   
5844                   0                 0                       0   
5845                   0                 1                       0   
5846                   0                 0                       0   

      Transmission_Manual  
0                       1  
1                       1  
2                       1  
3                       0  
4                       1  
...                   ...  
5842                    1  
5843                    1  
5844                    1  
5845                    1  
5846                    1  

[5847 rows x 15 columns]

In [ ]:
# Calculate the current year
current_year = datetime.datetime.now().year

# Create a new column "Car_Age" by subtracting "Year" from the current year
df['Car_Age'] = current_year - df['Year']

In [ ]:
df


Name    Location  Year  Kilometers_Driven  \
0     Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                         Honda Jazz V     Chennai  2011              46000   
2                    Maruti Ertiga VDI     Chennai  2012              87000   
3      Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4               Nissan Micra Diesel XV      Jaipur  2013              86999   
...                                ...         ...   ...                ...   
5842                  Maruti Swift VDI       Delhi  2014              27365   
5843          Hyundai Xcent 1.1 CRDi S      Jaipur  2015             100000   
5844             Mahindra Xylo D4 BSIV      Jaipur  2012              55000   
5845                Maruti Wagon R VXI     Kolkata  2013              46000   
5846             Chevrolet Beat Diesel   Hyderabad  2011              47000   

     Owner_Type  Mileage  Engine   Power  Seats  Price  Fuel_Type_Diesel  \
0         First    19.67  1582.0  126.20    5.0  12.50                 1   
1         First    13.00  1199.0   88.70    5.0   4.50                 0   
2         First    20.77  1248.0   88.76    7.0   6.00                 1   
3        Second    15.20  1968.0  140.80    5.0  17.74                 1   
4         First    23.08  1461.0   63.10    5.0   3.50                 1   
...         ...      ...     ...     ...    ...    ...               ...   
5842      First    28.40  1248.0   74.00    5.0   4.75                 1   
5843      First    24.40  1120.0   71.00    5.0   4.00                 1   
5844     Second    14.00  2498.0  112.00    8.0   2.90                 1   
5845      First    18.90   998.0   67.10    5.0   2.65                 0   
5846      First    25.44   936.0   57.60    5.0   2.50                 1   

      Fuel_Type_Electric  Fuel_Type_Petrol  Transmission_Automatic  \
0                      0                 0                       0   
1                      0                 1                       0   
2                      0                 0                       0   
3                      0                 0                       1   
4                      0                 0                       0   
...                  ...               ...                     ...   
5842                   0                 0                       0   
5843                   0                 0                       0   
5844                   0                 0                       0   
5845                   0                 1                       0   
5846                   0                 0                       0   

      Transmission_Manual  Car_Age  
0                       1        8  
1                       1       12  
2                       1       11  
3                       0       10  
4                       1       10  
...                   ...      ...  
5842                    1        9  
5843                    1        8  
5844                    1       11  
5845                    1       10  
5846                    1       12  

[5847 rows x 16 columns]

In [ ]:
df.to_csv('train_analysis.csv', index = False)